js/digifinex.js

```diff
@@ -0,0 +1,535 @@
+        };
+        if (since) {
+            request['date'] = this.dateUTC8 (since);
+        }
+        let response = await this.privateGetOrderHistory (params);
+        let filtered = this.parseOrders (response['orders'], undefined, since, limit);
+        return filtered;
+    }
+
+    async fetchOrder (id, symbol = undefined, params = {}) {
+        let response = await this.privateGetOrderInfo (this.extend ({
+            'order_id': parseInt (id),
+        }, params));
+        return this.parseOrder (response);
+    }
+
+    parseOHLCV (ohlcv, market = undefined, timeframe = '1m', since = undefined, limit = undefined) {
+        return [
+            ohlcv[0],
+            ohlcv[5],
+            ohlcv[3],
+            ohlcv[4],
+            ohlcv[2],
+            ohlcv[1],
+        ];
+    }
+
+    async fetchOHLCV (symbol, timeframe = '1m', since = undefined, limit = undefined, params = {}) {
+        let request = {
+            'symbol': symbol,
+            'type': 'kline_' + timeframe,
+        };
+        let response = await this.privateGetKline (this.extend (request, params));
+        return this.parseOHLCVs (response['data'], undefined, timeframe, since, limit);
+    }
+
+    nonce () {
+        return this.milliseconds ();
+    }
+
+    sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        params['apiKey'] = this.apiKey;
+        if (api === 'private') {
+            this.checkRequiredCredentials ();
+            params['timestamp'] = parseInt (this.seconds ());
+            params['apiSecret'] = this.secret;
+            params = this.keysort (params);
+            let keys = Object.keys (params);
+            let arr = [];
+            for (let i = 0; i < keys.length; i++) {
+                let key = keys[i];
+                let s = params[key].toString ();
+                arr.push (s);
+            }
+            let s = arr.join ('');
+            let sign = this.hash (this.encode (s), 'md5');
+            params['sign'] = sign;
+        }
+        let url = this.urls['api'] + '/' + path;
+        // log.info (url, params);
+        if (method === 'GET') {
+            if (Object.keys (params).length) {
+                url += '?' + this.urlencode (params);
+            }
+        } else if (method === 'POST') {
+            if (Object.keys (params).length) {
+                body = this.urlencode (params);
+            }
+        }
+        let result = { 'url': url, 'method': method, 'body': body, 'headers': headers };
+        // log.info (url);
+        // log.info (result);
+        return result;
+    }
+
+    dateUTC8 (timestampMS) {
+        const timedelta = this.safeValue (this.options, 'timedelta', 8 * 60 * 60 * 1000); // eight hours
+        return this.ymd (timestampMS + timedelta);
+    }
+
+    handleErrors (statusCode, statusText, url, method, responseHeaders, responseBody, response) {
+        this.checkResponse (response, '');
```

Comment:
This is getting better, now I suggest that we don't have `handleErrors` method just for calling another method – instead the body of `checkResponse` should be in `handleErrors`, and I suggest that we rewrite `checkCode` to use an `.exceptions` map like it's done in liqui or bitfinex.

Reply:
I don't think .exceptions map is good here. A switch-like if-else group is more straight forward.

Reply:
> A switch-like if-else group is more straight forward.

Nope, unfortunately, it's not, and it's much less convenient for the user – if a new exception code or message is added or changed – you could handle this via `.exceptions` without having to edit the code inside the method and rebuild the lib again. Basically the `.exceptions` mechanism is the standard way of doing it in CCXT, and if you see a different type of handling with if-else – that's mostly for historical reasons and to be replaced for `.exceptions`. So, in this case, and for most of new integrations, I suggest to stick to the unified approach, as it is done in other places throughout the lib. Most users will be really thankful for that. Let me know if you agree to the above )

Reply:
The error response of digifinex has only error code, without description message attached. If they add another error code, I must add a description myself. And it is more convenient to edit only the code, but not both the code and the map.

Reply:
The error code is most likely not to change, only new codes will be added. And most users don't have to distinguish different exceptions at all.

Reply:
> The error response of digifinex has only error code, without description message attached.

If the exchange does not provide an error message – I wouldn't add the text description in the first place we often return just the body of the response as is.

However, in some cases, if the user wants them, we have the mapping of exception codes to exception messages, and it's in the `describe()` as well, so if it comes to editing for a new message – the edit is still done in just one place (in the mapping inside `describe()`). We absolutely should not have the codes in one place and the messages in the other place. 

Here's what I'd ask you to do:

```JavaScript
    describe () {
        // ...
            'exceptions': {
                'exact': {
                    '10001': [ BadRequest, "Wrong request method, please check it's a GET ot POST request" ],
                    '10002': [ AuthenticationError, 'Invalid ApiKey' ],
                    '10003': [ AuthenticationError, "Sign doesn't match" ],
                    '10004': [ ArgumentsRequired, 'Illegal request parameters' ],
                    '10005': [ DDoSProtection, 'Request frequency exceeds the limit' ],
                    '10006': [ PermissionDenied, 'Unauthorized to execute this request' ],
                    '10007': [ PermissionDenied, 'IP address Unauthorized' ],
                    '10008': [ InvalidNonce, 'Timestamp for this request is invalid, timestamp must within 1 minute' ],
                    '10009': [ ExchangeError, 'Unexist endpoint, please check endpoint URL' ],
                    '10011': [ AccountSuspended, 'ApiKey expired. Please go to client side to re-create an ApiKey' ],
                    '20001': [ ExchangeError, 'Trade is not open for this trading pair' ],
                    '20002': [ ExchangeError, 'Trade of this trading pair is suspended' ],
                    '20003': [ InvalidOrder, 'Invalid price or amount' ],
                    '20007': [ InvalidOrder, 'Price precision error' ],
                    '20008': [ InvalidOrder, 'Amount precision error' ],
                    '20009': [ InvalidOrder, 'Amount is less than the minimum requirement' ],
                    '20010': [ InvalidOrder, 'Cash Amount is less than the minimum requirement' ],
                    '20011': [ InsufficientFunds, 'Insufficient balance' ],
                    '20012': [ BadRequest, 'Invalid trade type (valid value: buy/sell)' ],
                    '20013': [ OrderNotFound, 'No order info found' ],
                    '20014': [ BadRequest, 'Invalid date (Valid format: 2018-07-25)' ],
                    '20015': [ ExchangeError, 'Date exceeds the limit' ],
                    '20018': [ PermissionDenied, 'Your trading rights have been banned by the system' ],
                    '20019': [ BadRequest, 'Wrong trading pair symbol. Correct format:"usdt_btc". Quote asset is in the front' ],
                },
            },
        // ...
    }

    handleErrors (statusCode, statusText, url, method, responseHeaders, responseBody, response) {
        if (!response) {
            return; // fall back to default error handler
        }
        const code = this.safeString (response, 'code');
        if (code === '0') {
            return; // no error
        }
        if (code === undefined) {
            throw new BadResponse (this.id + ' ' + responseBody);
        }
        const feedback = this.id + ' ' + responseBody;
        const unknownError = [ ExchangeError, feedback ];
        const [ ExceptionClass, message ] = this.safeValue (this.exceptions['exact'], code, unknownError);
        throw new ExceptionClass (message);
    };
```

@szmcdull ↑ it's that easy and I think it is still better than having everything hardcoded, do you agree? )

Reply:
> please check it's a GET ot POST request

↑ We usually don't copy their existing messages because their English sucks, btw.

Reply:
hahaha... OK. Can a key be a number, instead of a string?

Reply:
> Can a key be a number, instead of a string?

It doesn't have to be anything but a string, and I've shown in my snippet above how the number-to-string conversion is done in `handleErrors()` in this line:

```JavaScript
        const code = this.safeString (response, 'code'); // safeString here instead of safeInteger
```

Let me know if that doesn't answer your question )